Source/MQTTnet/Client/MqttClient.cs

```diff
@@ -626,15 +628,25 @@ private async Task<bool> HandleReceivedApplicationMessageAsync(MqttPublishPacket
             return true;
         }
 
-        private static async Task WaitForTaskAsync(Task task, Task sender)
+        private async Task WaitForTaskAsync(Task task, Task sender)
         {
-            if (task == sender || task == null)
+            if (task == null)
             {
                 return;
             }
 
-            if (task.IsCanceled || task.IsCompleted || task.IsFaulted)
+            if (task == sender)
             {
+                // Return here to avoid deadlocks, but first any eventual exception in the task
+                // must be handled to avoid not getting an unhandled task exception
+                if (!task.IsFaulted)
+                {
+                    return;
+                }
+
+                // By accessing the Exception property the exception is considered handled and will
+                // not result in an unhandled task exception later by the finalizer
+                _logger.Warning(task.Exception, "Exception when waiting for background task.");
```

Comment:
Usually the errors messages start with "Error while...". Just a very minor thing but it should be aligned in my opinion.

Reply:
I agree.